client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx

```diff
@@ -118,22 +130,65 @@ const getAvailableUsers = (
   groups,
   group,
   hideInGroup,
+  hidePhantomStudent,
   availableSearch,
 ) => {
+  let groupMemberIds = new Set()
+  // const altGroup = new Set(groups.flatMap((g) => g.members).filter((m) => !m.isPhantom));
   if (hideInGroup) {
-    const allGroupMemberIds = new Set(
-      groups.flatMap((g) => g.members.map((m) => m.id)),
-    );
-    return filterByName(
-      availableSearch,
-      courseUsers.filter((cu) => !allGroupMemberIds.has(cu.id)),
-    );
+    groupMemberIds = new Set(groups.flatMap((g) => g.members.map((m) => m.id)));
+  } else {
+    groupMemberIds = new Set(group.members.map((m) => m.id))
   }
-  const groupMemberIds = new Set(group.members.map((m) => m.id));
-  return filterByName(
-    availableSearch,
+
+  const filteredGroup = filterByName(availableSearch,
     courseUsers.filter((cu) => !groupMemberIds.has(cu.id)),
   );
+
+  if (hidePhantomStudent) {
+    return filteredGroup.filter((m) => !m.isPhantom);
+  } 
+  return filteredGroup
+};
+
+const getSelectedUsers = (
+  members,
+  selectedSearch,
+  hidePhantomStudent,
+) => {
+  let groupMembers = new Set()
+  if (hidePhantomStudent) {
+    groupMembers = new Set(members.filter((m) => !m.isPhantom))
+  } else {
+    groupMembers = new Set(members)
+  }
+
+  return filterByName(
+    selectedSearch,
+    [...groupMembers],
+  )
+};
+
+const getAvailableUserInOtherGroups = (
```

Comment:
Actually I suggest this to be done on the backend side.
Maybe `json.otherGroupIds` to be added to `json.members` in `coursemology2/app/views/course/group/_group.json.jbuilder`. This way, we can reduce the complexity in the frontend and let the backend do the heavy lifting. Also, since there's no other interaction that will cause changes to the frontend view dynamically, it makes sense to do it in the backend. (And most importantly for you to work on some backend code)